lib/state_sampler.h

```diff
@@ -0,0 +1,56 @@
+#ifndef STATE_SAMPLER_H_
+#define STATE_SAMPLER_H_
+
+#include <complex>
+#include <memory>
+#include <random>
+
+namespace qsim {
+
+template <typename StateSpace>
+class StateSampler {
+ public:
+  using State = typename StateSpace::State;
+
+  void SampleState(const StateSpace& state_space, const State& state,
+                   const unsigned num_samples, unsigned seed,
+                   std::vector<uint64_t>* samples) {
+    if (num_samples <= 0) {
+      return;
+    }
+    samples->reserve(num_samples);
+
+    double norm = 0;
+    for (uint64_t k = 0; k < state_space.Size(state); ++k) {
+      norm += std::norm(state_space.GetAmpl(state, k));
+    }
+
+    std::mt19937_64 rgen(seed);
+    double rgen_norm = norm / std::mt19937_64::max();
```

Comment:
It is probably better to use `std::uniform_real_distribution(0.0, norm)` as in your initial commit.